sklearn/cluster/_hdbscan/_linkage.pyx

```diff
@@ -14,197 +16,180 @@ from ...cluster._hierarchical_fast cimport UnionFind
+        cnp.float64_t next_node_min_reach, pair_distance, next_node_core_dist
+
+    n_samples = raw_data.shape[0]
     num_features = raw_data.shape[1]
 
-    raw_data_view = (<cnp.double_t[:raw_data.shape[0], :raw_data.shape[1]:1]> (
-        <cnp.double_t *> raw_data.data))
-    raw_data_ptr = (<cnp.double_t *> &raw_data_view[0, 0])
+    mst = np.empty(n_samples - 1, dtype=MST_edge_dtype)
 
-    result_arr = np.zeros((dim - 1, 3))
-    in_tree_arr = np.zeros(dim, dtype=np.int8)
-    current_node = 0
-    current_distances_arr = np.infty * np.ones(dim)
-    current_sources_arr = np.ones(dim)
+    in_tree = np.zeros(n_samples, dtype=np.int8)
+    min_reachability = np.infty * np.ones(n_samples)
+    current_sources = np.ones(n_samples)
 
-    result = (<cnp.double_t[:dim - 1, :3:1]> (<cnp.double_t *> result_arr.data))
-    in_tree = (<cnp.int8_t *> in_tree_arr.data)
-    current_distances = (<cnp.double_t *> current_distances_arr.data)
-    current_sources = (<cnp.double_t *> current_sources_arr.data)
-    current_core_distances = (<cnp.double_t *> core_distances.data)
+    current_node = 0
 
-    for i in range(1, dim):
+    for i in range(1, n_samples):
 
         in_tree[current_node] = 1
 
-        current_node_core_distance = current_core_distances[current_node]
+        current_node_core_dist = current_core_distances[current_node]
 
-        new_distance = DBL_MAX
+        new_reachability = DBL_MAX
         source_node = 0
         new_node = 0
 
-        for j in range(dim):
+        for j in range(n_samples):
             if in_tree[j]:
                 continue
 
-            right_value = current_distances[j]
-            right_source = current_sources[j]
+            next_node_min_reach = min_reachability[j]
+            next_node_source = current_sources[j]
 
-            left_value = dist_metric.dist(&raw_data_ptr[num_features *
-                                                        current_node],
-                                          &raw_data_ptr[num_features * j],
-                                          num_features)
-            left_source = current_node
+            pair_distance = dist_metric.dist(
+                &raw_data_view[current_node, 0],
+                &raw_data_view[j, 0],
+                num_features
+            )
 
             if alpha != 1.0:
-                left_value /= alpha
-
-            core_value = core_distances[j]
-            if (current_node_core_distance > right_value or
-                    core_value > right_value or
-                    left_value > right_value):
-                if right_value < new_distance:
-                    new_distance = right_value
-                    source_node = right_source
+                pair_distance /= alpha
+
+            next_node_core_dist = core_distances[j]
+            mutual_reachability_distance = max(
+                current_node_core_dist,
+                next_node_core_dist,
+                pair_distance
+            )
```

Comment:
As there is no import-shadowing, `max` likely is the built-in Python function.
I think we might prefer using `libc.math.fmax` twice instead.

Reply:
`max` is a Cython built-in, is it not? At least the annotated Cython shows no Python interaction at all.

Reply:
Thanks for the information, probably Cython convert `max` to use something efficient under the hood.